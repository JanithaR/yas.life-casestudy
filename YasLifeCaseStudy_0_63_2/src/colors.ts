export default {
    background: '#e7e4fb',
    primary: '#5c53af',
};
