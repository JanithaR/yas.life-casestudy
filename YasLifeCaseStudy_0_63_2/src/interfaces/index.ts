export * from './Currency';
export * from './FixerRate';
export * from './FixerLatest';
