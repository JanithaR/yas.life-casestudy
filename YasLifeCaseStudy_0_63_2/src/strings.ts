export default {
    inputPlaceholder: 'Enter value',
    pickerPlaceholder: 'Choose currency',
    convert: 'Convert',
    clickConvert: 'Click Convert',
    invalidInput: 'Please enter a valid input',
};
