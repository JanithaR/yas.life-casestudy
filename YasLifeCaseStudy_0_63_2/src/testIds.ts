export default {
    picker: 'picker',
    activityIndicator: 'activityIndicator',
};
