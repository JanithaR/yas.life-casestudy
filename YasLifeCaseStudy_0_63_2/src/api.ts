import { FixerLatest } from 'src/interfaces';

export function callApi(url: string): Promise<FixerLatest> {
    return fetch(url).then((response) => response.json());
}
