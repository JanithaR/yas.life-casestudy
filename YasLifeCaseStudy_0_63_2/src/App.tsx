/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * Generated with the TypeScript template
 * https://github.com/react-native-community/react-native-template-typescript
 *
 * @format
 */

import React, { ReactElement } from 'react';
import {
    SafeAreaView,
    StyleSheet,
    ScrollView,
    StatusBar,
    TextInput,
    Button,
    View,
    Text,
    ActivityIndicator,
} from 'react-native';

import { Picker } from '@react-native-community/picker';

import strings from 'src/strings';
import colors from 'src/colors';
import testIds from 'src/testIds';
import {
    formatCurrency,
    convertCurrency,
    composeLatestEndpointUrl,
} from 'src/utils';
import { currencies } from 'src/config';
import { callApi } from 'src/api';
import { FixerLatest, Rate } from 'src/interfaces';

declare const global: { HermesInternal: null | {} };

const App = () => {
    const [currency, setCurrency] = React.useState<string>(currencies[1].code);
    const [input, setInput] = React.useState<number>(1);
    const [rates, setRates] = React.useState<Rate>({});
    const [loading, setLoading] = React.useState<boolean>(false);

    function onCurrencyChange(value: string | number): void {
        let validatedCurrency: string;

        if (typeof value === 'number') {
            validatedCurrency = value.toString();
        } else {
            validatedCurrency = value;
        }

        setCurrency(validatedCurrency);
    }

    function onConvertPress(): void {
        setLoading(true);

        callApi(composeLatestEndpointUrl()).then((data: FixerLatest) => {
            setLoading(false);
            setRates(data.rates);
        });
    }

    function onInputChange(value: string): void {
        const parsedFloat: number = parseFloat(value);

        if (!isNaN(parsedFloat)) {
            setInput(parsedFloat);
        } else {
            setInput(0);
        }
    }

    function getConversionRate(curr: string): Rate['key'] {
        return rates[curr];
    }

    function renderConversion(): ReactElement | null {
        return Object.keys(rates).length > 0 ? (
            <Text style={styles.outputText}>
                {formatCurrency(
                    convertCurrency(input, getConversionRate(currency)),
                    currency,
                )}
            </Text>
        ) : null;
    }

    function renderPickerItems(): ReactElement[] {
        return currencies
            .filter((curr, index) => index !== 0)
            .map((curr) => (
                <Picker.Item
                    label={curr.label}
                    value={curr.code}
                    key={curr.code}
                />
            ));
    }

    function renderConvertButton(): ReactElement {
        return loading ? (
            <ActivityIndicator
                size="large"
                color={colors.primary}
                testID={testIds.activityIndicator}
            />
        ) : (
            <Button
                title={strings.convert}
                onPress={onConvertPress}
                color={colors.primary}
            />
        );
    }

    function renderInstructions(): ReactElement | null {
        return !loading && Object.keys(rates).length === 0 ? (
            <Text>{strings.clickConvert}</Text>
        ) : null;
    }

    return (
        <>
            <StatusBar barStyle="light-content" />
            <SafeAreaView style={styles.safeArea}>
                <ScrollView
                    contentInsetAdjustmentBehavior="automatic"
                    style={styles.scrollView}>
                    <View style={styles.sectionWrapper}>
                        <TextInput
                            value={input > 0 ? input.toString() : ''}
                            onChangeText={onInputChange}
                            keyboardType="decimal-pad"
                            selectTextOnFocus
                            style={styles.input}
                            placeholder={strings.inputPlaceholder}
                        />
                        <Picker
                            testID={testIds.picker}
                            selectedValue={currency}
                            onValueChange={onCurrencyChange}
                            itemStyle={styles.picker}>
                            {renderPickerItems()}
                        </Picker>
                        <View style={styles.buttonWrapper}>
                            {renderConvertButton()}
                        </View>
                    </View>
                    <View
                        style={StyleSheet.flatten([
                            styles.sectionWrapper,
                            styles.outputWrapper,
                        ])}>
                        <Text style={styles.inputText}>
                            {formatCurrency(input)} =
                        </Text>

                        {renderConversion()}

                        {renderInstructions()}
                    </View>
                </ScrollView>
            </SafeAreaView>
        </>
    );
};

const styles = StyleSheet.create({
    scrollView: {
        backgroundColor: colors.background,
        padding: 20,
    },
    sectionWrapper: {
        backgroundColor: 'white',
        borderRadius: 10,
        padding: 10,
    },
    input: {
        borderRadius: 10,
        borderColor: colors.background,
        borderWidth: 1,
        fontSize: 36,
    },
    safeArea: {
        flex: 1,
    },
    outputWrapper: {
        marginTop: 10,
        alignItems: 'center',
    },
    picker: {
        fontSize: 26,
    },
    inputText: {
        color: colors.primary,
        fontSize: 36,
    },
    buttonWrapper: {
        marginTop: 10,
    },
    outputText: {
        color: colors.primary,
        fontSize: 56,
    },
});

export default App;
