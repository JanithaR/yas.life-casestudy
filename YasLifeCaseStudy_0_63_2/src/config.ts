import { Currency } from 'src/interfaces';

export const baseApiUrl: string = 'http://data.fixer.io/api/';
export const latestEndpoint: string = 'latest';
export const currencies: Currency[] = [
    { label: 'Europe', code: 'EUR' },
    { label: 'Sri Lanka', code: 'LKR' },
    { label: 'US', code: 'USD' },
    { label: 'United Arab Emirates', code: 'AED' },
];
