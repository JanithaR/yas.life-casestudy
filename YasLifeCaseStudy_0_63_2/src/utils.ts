import { Currency } from 'src/interfaces';
import { currencies, baseApiUrl, latestEndpoint, fixerKey } from 'src/config';

export function formatCurrency(
    value: number,
    currency: Currency['code'] = currencies[0].code,
): string {
    return new Intl.NumberFormat('en-US', {
        style: 'currency',
        currency,
        maximumFractionDigits: 2,
    })
        .format(value)
        .toString();
}

export function convertCurrency(value: number, conversionRate: number): number {
    return value * conversionRate;
}

export const getCommaSeparatedCurrencyCodes = (): string => {
    let output = '';

    currencies
        .filter((currency, index) => index !== 0)
        .forEach((currency) => {
            output = `${output}${currency.code},`;
        });

    return output;
};

export function composeLatestEndpointUrl(): string {
    return `${baseApiUrl}${latestEndpoint}?access_key=${fixerKey}&symbols=${getCommaSeparatedCurrencyCodes()}`;
}
