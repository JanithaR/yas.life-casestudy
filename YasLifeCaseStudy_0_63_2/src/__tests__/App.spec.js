import React from 'react';
import { render, fireEvent, waitFor } from 'react-native-testing-library';
import {
    // eslint-disable-next-line no-unused-vars
    toBeEnabled,
    // eslint-disable-next-line no-unused-vars
    toHaveTextContent,
    // eslint-disable-next-line no-unused-vars
    toHaveProp,
} from '@testing-library/jest-native';
import AsyncStorage from '@react-native-community/async-storage';

import App from 'src/App';
import strings from 'src/strings';
import testIds from 'src/testIds';
import { convertCurrency, formatCurrency } from 'src/utils';
import { currencies } from 'src/config';
import { callApi } from 'src/api';

jest.mock('src/api', () => ({
    callApi: jest.fn(() =>
        Promise.resolve({
            base: 'EUR',
            date: '2020-07-27',
            rates: {
                DKK: 7.443104,
                INR: 87.741498,
                LKR: 218.134047,
                THB: 37.007496,
                USD: 1.174102,
            },
            success: true,
            timestamp: 1595851446,
        }),
    ),
}));

function setup(): RenderAPI {
    return render(<App />);
}

describe('App screen', () => {
    it('should render as expected on launch', () => {
        const { toJSON } = setup();

        expect(toJSON()).toMatchSnapshot();
    });

    it('should render as expected when an input is converted', async () => {
        const { queryByA11yRole, toJSON } = setup();
        const button = queryByA11yRole('button');

        fireEvent(button, 'onPress');

        await waitFor(() => button);

        expect(toJSON()).toMatchSnapshot();
    });

    describe('Input', () => {
        beforeEach(() => {
            AsyncStorage.setItem.mockClear();
        });

        it('should be rendered', () => {
            const { queryByDisplayValue } = setup();
            const input = queryByDisplayValue('1');

            expect(input).toBeTruthy();
            expect(input).toBeEnabled();
        });

        it('should show 1 as default value', () => {
            const { queryByDisplayValue } = setup();

            expect(queryByDisplayValue('1')).toBeTruthy();
        });

        it('should be pre populated if a cached input exists', async () => {
            AsyncStorage.getItem.mockResolvedValueOnce('1234.56');

            const { queryByDisplayValue } = setup();

            await waitFor(() => queryByDisplayValue('1234.56'));

            expect(queryByDisplayValue('1234.56')).toBeTruthy();
        });

        it('should show default value 1 if an exception occurred when reading cached input', async () => {
            AsyncStorage.getItem.mockRejectedValueOnce({
                message: 'Async storage read exception',
            });

            const { queryByText } = setup();

            await waitFor(() => {
                queryByText('Async storage read exception');
            });

            expect(queryByText('Async storage read exception')).toBeTruthy();
        });

        it('should update the input field and cache the value when typed', () => {
            const { queryByDisplayValue } = setup();

            fireEvent(queryByDisplayValue('1'), 'onChangeText', '2');

            expect(queryByDisplayValue('2')).toBeTruthy();

            expect(AsyncStorage.setItem).toHaveBeenCalledTimes(1);
            expect(AsyncStorage.setItem).toBeCalledWith('input', '2');

            fireEvent(queryByDisplayValue('2'), 'onChangeText', '3.5');

            expect(queryByDisplayValue('3.5')).toBeTruthy();

            expect(AsyncStorage.setItem).toHaveBeenCalledTimes(2);
            expect(AsyncStorage.setItem).toBeCalledWith('input', '3.5');
        });

        it('should show an error if writing cache failed', async () => {
            AsyncStorage.setItem.mockRejectedValueOnce({
                message: 'Async storage write exception',
            });

            const { queryByDisplayValue, queryByText } = setup();

            fireEvent(queryByDisplayValue('1'), 'onChangeText', '2');

            await waitFor(() => {
                queryByText('Async storage write exception');
            });

            expect(queryByText('Async storage write exception')).toBeTruthy();
        });

        it('should instruct the user to input a value if it is empty', () => {
            const { queryByDisplayValue, queryByPlaceholder } = setup();

            fireEvent(queryByDisplayValue('1'), 'onChangeText', '');

            expect(queryByPlaceholder(strings.inputPlaceholder)).toBeTruthy();
        });
    });

    describe('Currency picker', () => {
        it('should be renderd', () => {
            // no other way to test this?
            const { queryByTestId } = setup();
            const picker = queryByTestId(testIds.picker);

            expect(picker).toBeTruthy();
            expect(picker).toBeEnabled();
        });

        it('should change the selected index when changed', () => {
            const { getByTestId } = setup();
            const picker = getByTestId(testIds.picker);

            expect(picker).toHaveProp('selectedIndex', 0);

            fireEvent(picker, 'onValueChange', 'LKR');

            expect(picker).toHaveProp('selectedIndex', 3);
        });
    });

    describe('Convert button', () => {
        beforeEach(() => {
            callApi.mockClear();
        });

        it('should be rendered', () => {
            const { queryByA11yRole } = setup();
            const button = queryByA11yRole('button');

            expect(button).toBeEnabled();
            expect(button).toHaveTextContent(strings.convert);
        });

        it('should call callApi() from api utils when pressed', () => {
            const { queryByA11yRole } = setup();

            fireEvent(queryByA11yRole('button'), 'onPress');

            expect(callApi).toHaveBeenCalledTimes(1);

            expect(callApi).toHaveBeenCalledWith(
                expect.stringMatching(
                    /^http:\/\/data.fixer.io\/api\/latest\?access_key=[a-z0-9]{32}&symbols=.+$/,
                ),
            );
        });

        it('should not be rendered when fetching rates, instead an activity indicator should be rendered', async () => {
            const { queryByA11yRole, getByTestId } = setup();

            fireEvent(queryByA11yRole('button'), 'onPress');

            expect(queryByA11yRole('button')).toBeFalsy();
            expect(getByTestId(testIds.activityIndicator)).toBeTruthy();
        });
    });

    describe('Output', () => {
        beforeEach(() => {
            AsyncStorage.getItem.mockImplementationOnce((key: string) =>
                Promise.resolve(null),
            );
        });

        it('should show the input value in base currency', () => {
            const { queryByText } = setup();

            expect(
                queryByText(`${formatCurrency(1, currencies[0].code)} =`),
            ).toBeTruthy();
        });

        it('should not show the output value on launch', () => {
            const { queryByText } = setup();
            const expectedOutput: string = formatCurrency(
                convertCurrency(1, 1),
                currencies[1].code,
            );

            expect(queryByText(expectedOutput)).toBeFalsy();
        });

        it('should update the base currency value when input value changes', () => {
            const { getByDisplayValue, queryByText } = setup();
            const input = getByDisplayValue('1');

            fireEvent(input, 'onChangeText', '2');

            expect(
                queryByText(`${formatCurrency(2, currencies[0].code)} =`),
            ).toBeTruthy();
        });

        it('should instruct the user to click Convert on launch', () => {
            const { queryByText } = setup();

            expect(queryByText(strings.clickConvert)).toBeTruthy();
        });

        it('should show latest rates fetching status upon clicking Convert', async () => {
            const { getByA11yRole, queryByText } = setup();

            fireEvent(getByA11yRole('button'), 'onPress');

            await waitFor(() => getByA11yRole('button'));

            expect(queryByText(strings.fetchingLatestRates)).toBeFalsy();
        });

        it('should show converted value in requested currency', async () => {
            const { getByA11yRole, queryByText } = setup();

            fireEvent(getByA11yRole('button'), 'onPress');

            await waitFor(() => getByA11yRole('button'));

            const apiResult: FixerLatest = await callApi();

            expect(
                queryByText(
                    formatCurrency(
                        convertCurrency(1, apiResult.rates[currencies[1].code]),
                        currencies[1].code,
                    ),
                ),
            ).toBeTruthy();
        });

        it('should show error description if API threw an error', async () => {
            callApi.mockResolvedValueOnce({
                success: false,
                error: {
                    code: 104,
                    info: 'Fixer API error description',
                },
            });

            const { queryByText, getByA11yRole } = setup();

            fireEvent(getByA11yRole('button'), 'onPress');

            await waitFor(() => getByA11yRole('button'));

            expect(queryByText('Fixer API error description')).toBeTruthy();
        });

        it('should show error description if an API exception occurred', async () => {
            callApi.mockRejectedValueOnce({
                message: 'API exception',
            });

            const { queryByText, getByA11yRole } = setup();

            fireEvent(getByA11yRole('button'), 'onPress');

            await waitFor(() => getByA11yRole('button'));

            expect(queryByText('API exception')).toBeTruthy();
        });

        it('should update the converted value when currency changes', async () => {
            const { getByA11yRole, queryByText, getByTestId } = setup();

            fireEvent(getByA11yRole('button'), 'onPress');

            await waitFor(() => getByA11yRole('button'));

            fireEvent(
                getByTestId(testIds.picker),
                'onValueChange',
                currencies[2].code,
            );

            const apiResult: FixerLatest = await callApi();

            expect(
                queryByText(
                    formatCurrency(
                        convertCurrency(1, apiResult.rates[currencies[2].code]),
                        currencies[2].code,
                    ),
                ),
            ).toBeTruthy();

            fireEvent(
                getByTestId(testIds.picker),
                'onValueChange',
                currencies[2].code,
            );

            expect(
                queryByText(
                    formatCurrency(
                        convertCurrency(1, apiResult.rates[currencies[2].code]),
                        currencies[2].code,
                    ),
                ),
            ).toBeTruthy();
        });

        it('should update the converted value when input changes', async () => {
            const { getByA11yRole, queryByText, getByDisplayValue } = setup();

            fireEvent(getByA11yRole('button'), 'onPress');

            await waitFor(() => getByA11yRole('button'));

            fireEvent(getByDisplayValue('1'), 'onChangeText', '563');

            const apiResult: FixerLatest = await callApi();

            expect(
                queryByText(
                    formatCurrency(
                        convertCurrency(
                            563,
                            apiResult.rates[currencies[1].code],
                        ),
                        currencies[1].code,
                    ),
                ),
            ).toBeTruthy();

            fireEvent(getByDisplayValue('563'), 'onChangeText', '463');

            expect(
                queryByText(
                    formatCurrency(
                        convertCurrency(
                            463,
                            apiResult.rates[currencies[1].code],
                        ),
                        currencies[1].code,
                    ),
                ),
            ).toBeTruthy();
        });

        it('should instruct the user to enter a valid input in case it is invalid', () => {
            const { getByDisplayValue, queryByText } = setup();

            fireEvent(getByDisplayValue('1'), 'onChangeText', '');

            expect(queryByText(strings.invalidInput)).toBeTruthy();

            let currencySymbol = formatCurrency(
                Number.parseFloat(''),
                currencies[0].code,
            ).replace('NaN', '');

            let regex = new RegExp(currencySymbol);
            expect(queryByText(regex)).toBeFalsy();

            currencySymbol = formatCurrency(
                Number.parseFloat(''),
                currencies[1].code,
            ).replace('NaN', '');

            regex = new RegExp(currencySymbol);
            expect(queryByText(regex)).toBeFalsy();
        });
    });
});
