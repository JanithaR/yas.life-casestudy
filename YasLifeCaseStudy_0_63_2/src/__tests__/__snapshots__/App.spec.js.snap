// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App screen should render as expected 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentInsetAdjustmentBehavior="automatic"
    style={
      Object {
        "backgroundColor": "#e7e4fb",
        "padding": 20,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 10,
            "padding": 10,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="decimal-pad"
          onChangeText={[Function]}
          placeholder="Enter value"
          rejectResponderTermination={true}
          selectTextOnFocus={true}
          style={
            Object {
              "borderColor": "#e7e4fb",
              "borderRadius": 10,
              "borderWidth": 1,
              "fontSize": 36,
            }
          }
          underlineColorAndroid="transparent"
          value="1"
        />
        <View>
          <RNCPicker
            items={
              Array [
                Object {
                  "label": "Sri Lanka",
                  "textColor": undefined,
                  "value": "LKR",
                },
                Object {
                  "label": "Danmark",
                  "textColor": undefined,
                  "value": "DKK",
                },
                Object {
                  "label": "India",
                  "textColor": undefined,
                  "value": "INR",
                },
                Object {
                  "label": "US",
                  "textColor": undefined,
                  "value": "USD",
                },
                Object {
                  "label": "Thailand",
                  "textColor": undefined,
                  "value": "THB",
                },
              ]
            }
            onChange={[Function]}
            selectedIndex={0}
            style={
              Array [
                Object {
                  "height": 216,
                },
                Object {
                  "fontSize": 26,
                },
              ]
            }
            testID="picker"
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={Object {}}
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#5c53af",
                    },
                  ]
                }
              >
                Convert
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 10,
            "marginTop": 10,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#5c53af",
              "fontSize": 36,
            }
          }
        >
          â‚¬1.00
           =
        </Text>
        <Text>
          Click Convert
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
